internal/thea.go

```diff
@@ -3,323 +3,196 @@ package internal
+
+	ctx, cancel := context.WithCancel(parent)
+	crashHandler := func(label string, err error) {
+		log.Emit(logger.FATAL, "Service crash (%s)! %s\n", label, err.Error())
+		cancel()
 	}
 
-	thea.spawnAsyncService(thea.workers.StartWorkers)
-	thea.spawnAsyncService(thea.ffmpegMgr.Start)
-
-	updateTicker := time.NewTicker(THEA_UPDATE_INTERVAL)
-	queueSyncTicker := time.NewTicker(THEA_QUEUE_SYNC_INTERVAL)
-
-	log.Emit(logger.SUCCESS, " --- Thea Startup Complete --- \n")
-
-	for {
-		select {
-		case <-updateTicker.C:
-			thea.SubmitUpdates()
-		case <-queueSyncTicker.C:
-			if err := thea.synchroniseQueue(); err != nil {
-				log.Emit(logger.WARNING, "Failed to synchronise item queue: %s\n", err.Error())
-			}
-		case <-exitChannel:
-			log.Emit(logger.STOP, "Interrupt detected!\n")
-			return nil
-		case <-thea.theaCtx.Done():
-			log.Emit(logger.WARNING, "Context has been cancelled!\n")
-			return nil
-		}
+	log.Emit(logger.NEW, "Initialising Docker services...\n")
+	if err := thea.initialiseDockerServices(thea.config, crashHandler); err != nil {
+		return err
 	}
-}
-
-// Stop will terminate Thea
-func (thea *theaImpl) Stop() {
-	log.Emit(logger.STOP, "--- Thea is shutting down ---\n")
-
-	log.Emit(logger.STOP, "Closing all managers...\n")
-	thea.workers.CloseWorkers()
-	thea.ffmpegMgr.Stop()
-	thea.shutdownWaitGroup.Wait()
-
-	log.Emit(logger.STOP, "Closing Docker containers...\n")
-	docker.DockerMgr.Shutdown(time.Second * 15)
-
-	log.Emit(logger.STOP, "Cancelling context...\n")
-	thea.theaCtxCancel()
-	log.Emit(logger.DEBUG, "Thea core shutdown complete\n")
-}
-
-// ** INTERNAL API ** //
-func (thea *theaImpl) queue() queue.QueueManager        { return thea.queueMgr }
-func (thea *theaImpl) ffmpeg() ffmpeg.FfmpegCommander   { return thea.ffmpegMgr }
-func (thea *theaImpl) profiles() profile.ProfileManager { return thea.profileMgr }
-func (thea *theaImpl) workerPool() *worker.WorkerPool   { return thea.workers }
-func (thea *theaImpl) config() TheaConfig               { return thea.cfg }
-
-// ** PRIVATE IMPL ** //
-
-// spawnAsyncService will run the provided function/service as it's own
-// go-routine, ensuring that the Thea service waitgroup is updated correctly
-func (thea *theaImpl) spawnAsyncService(service func()) {
-	thea.shutdownWaitGroup.Add(1)
 
-	go func() {
-		defer thea.shutdownWaitGroup.Done()
-		service()
-	}()
-}
-
-// synchroniseQueue will first discover all items inside the import directory,
-// and then will injest any that do not already exist in the queue. Any items
-// in the queue that no longer exist in the discovered items will also be cancelled
-func (thea *theaImpl) synchroniseQueue() error {
-	// Find new items
-	thea.queueMgr.Reload()
-	presentItems, err := thea.discoverItems()
-	if err != nil {
+	log.Emit(logger.NEW, "Connecting to database...\n")
+	db := database.New()
+	if store, err := NewStoreOrchestrator(db); err == nil {
+		thea.storeOrchestrator = store
+	} else {
 		return err
 	}
```

Comment:
Flip these. Always handle the error case first and then you can remove the `else`. The theory is that this puts the entire "happy path" at the same indentation level.

Reply:
Yeah, typically I do, but I didn't want to pollute the namespace of the function with `store` as I only wanted to store it in the `thea` receiver anyway. This is one of the few times I wish go had a littttttle bit of sugar around error handling.